File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/runtime/ValueSet.java
package org.opencds.cqf.cql.engine.runtime;

import java.util.ArrayList;
import java.util.List;

public class ValueSet extends Vocabulary {
    public ValueSet withId(String id) {
        setId(id);
        return this;
    }
    public ValueSet withVersion(String version) {
        setVersion(version);
        return this;
    }
    public ValueSet withName(String name) {
        setName(name);
        return this;
    }

    private List<CodeSystem> codeSystems = new ArrayList<CodeSystem>();
    public Iterable<CodeSystem> getCodeSystems() {
        return codeSystems;
    }
    public void setCodeSystems(List<CodeSystem> codeSystems) {
        this.codeSystems = new ArrayList<CodeSystem>();
        if (codeSystems != null) {
            for (CodeSystem cs : codeSystems) {
                if (cs != null) {
                    addCodeSystem(cs);
                }
            }
        }
    }
    public ValueSet withCodeSystems(List<CodeSystem> codeSystems) {
        setCodeSystems(codeSystems);
        return this;
    }
    public void addCodeSystem(CodeSystem codeSystem) {
        if (codeSystem == null) {
            throw new IllegalArgumentException("codeSystem is required");
        }
        codeSystems.add(codeSystem);
    }
    public ValueSet withCodeSystem(CodeSystem codeSystem) {
        addCodeSystem(codeSystem);
        return this;
    }
    public CodeSystem getCodeSystem(String id) {
        if (id == null) {
            return null;
        }

        for (CodeSystem cs : codeSystems) {
            if (id.equals(cs.getId())) {
                return cs;
            }
        }

        return null;
    }
    public CodeSystem getCodeSystem(String id, String version) {
        if (id == null) {
            return null;
        }

        for (CodeSystem cs : codeSystems) {
            if (id.equals(cs.getId()) && ((version == null && cs.getVersion() == null) || (version != null && version.equals(cs.getVersion())))) {
                return cs;
            }
        }

        return null;
    }

    @Override
    public Boolean equivalent(Object other) {
        if (!(other instanceof ValueSet)) {
            return false;
        }
        ValueSet otherV = (ValueSet)other;
        Boolean equivalent = super.equivalent(other) && codeSystems.size() == otherV.codeSystems.size();
        if (equivalent) {
            for (CodeSystem cs : codeSystems) {
                CodeSystem otherC = otherV.getCodeSystem(cs.getId());
                if (otherC == null) {
                    return false;
                }
            }
        }
        return equivalent;
    }

    @Override
    public Boolean equal(Object other) {
        if (!(other instanceof ValueSet)) {
            return false;
        }
        ValueSet otherV = (ValueSet)other;
        Boolean equal = super.equal(other) && codeSystems.size() == otherV.codeSystems.size();
        if (equal) {
            for (CodeSystem cs : codeSystems) {
                CodeSystem otherC = otherV.getCodeSystem(cs.getId(), cs.getVersion());
                if (otherC == null) {
                    return false;
                }
            }
        }
        return equal;
    }
}

File: Src/java/engine/src/main/kotlin/org/opencds/cqf/cql/engine/runtime/ValueSet.kt
package org.opencds.cqf.cql.engine.runtime

class ValueSet : Vocabulary() {
    fun withId(id: String?): ValueSet {
        this.id = id
        return this
    }

    fun withVersion(version: String?): ValueSet {
        this.version = version
        return this
    }

    fun withName(name: String?): ValueSet {
        this.name = name
        return this
    }

    private var codeSystems = mutableListOf<CodeSystem>()

    fun getCodeSystems(): Iterable<CodeSystem> {
        return codeSystems
    }

    fun setCodeSystems(codeSystems: MutableList<CodeSystem?>?) {
        this.codeSystems = mutableListOf()
        if (codeSystems != null) {
            for (cs in codeSystems) {
                if (cs != null) {
                    addCodeSystem(cs)
                }
            }
        }
    }

    fun withCodeSystems(codeSystems: MutableList<CodeSystem?>?): ValueSet {
        setCodeSystems(codeSystems)
        return this
    }

    fun addCodeSystem(codeSystem: CodeSystem) {
        codeSystems.add(codeSystem)
    }

    fun withCodeSystem(codeSystem: CodeSystem): ValueSet {
        addCodeSystem(codeSystem)
        return this
    }

    fun getCodeSystem(id: String?): CodeSystem? {
        if (id == null) {
            return null
        }

        for (cs in codeSystems) {
            if (id == cs.id) {
                return cs
            }
        }

        return null
    }

    fun getCodeSystem(id: String?, version: String?): CodeSystem? {
        if (id == null) {
            return null
        }

        for (cs in codeSystems) {
            if (
                id == cs.id &&
                    ((version == null && cs.version == null) ||
                        (version != null && version == cs.version))
            ) {
                return cs
            }
        }

        return null
    }

    override fun equivalent(other: Any?): Boolean? {
        if (other !is ValueSet) {
            return false
        }
        val otherV = other
        val equivalent =
            super.equivalent(other) == true && codeSystems.size == otherV.codeSystems.size
        if (equivalent) {
            for (cs in codeSystems) {
                val otherC = otherV.getCodeSystem(cs.id)
                if (otherC == null) {
                    return false
                }
            }
        }
        return equivalent
    }

    override fun equal(other: Any?): Boolean? {
        if (other !is ValueSet) {
            return false
        }
        val equal = super.equal(other) == true && codeSystems.size == other.codeSystems.size
        if (equal) {
            for (cs in codeSystems) {
                val otherC = other.getCodeSystem(cs.id, cs.version)
                if (otherC == null) {
                    return false
                }
            }
        }
        return equal
    }
}
